### PR TITLE
Implement help command for REPL NOSTR client

### PR DESCRIPTION
**Pull request by vladimirfomene**
Not a big change but I really wanted to get my feet wet.

### CODE REVIEW

File: src/sample.rs [hunk 1]
```diff
@@ -44,8 +44,24 @@ async fn poll_for_user_input(client_keys: Keys, tx: futures_channel::mpsc::Unbou
...
+					println!("    args: <content> - String of the text sent to the relay");
+					println!("  setmetadata\tSet the metadata for this client");
+					println!("    args: <username> <about> <picture> - The client's metadata strings to set");
+					println!("  recommendserver\tRecommend a server to the relay");
+					println!("    args: <urlrelay> - The url string of the server to recommend");
+					println!("  opensubscription\tOpen a subscription to the relay");
+					println!("    args: <subscriptionid> <kinds> <since> <until> - The metadata strings of the subscription to open");
+					println!("  closesubscription\tClose a subscription to the relay");
+					println!("    args: <subscriptionid> - The subscription to close");
+					println!("  shutdown\tShutdown the relay");
```
Comment (kouloumos) [inline src/sample.rs:61]: minor correction: this doesn't shutdown the relay, but instead closes the connection that the client (`civkit-sample`) has with the relay.

**Review by kouloumos — COMMENTED** [review]
Looks good for starts!

This is the preview of the new help text:
```
> help

Usage: <command> [args]


Available commands (See Nostr NIPs for more details):
  help  Show a list of available commands
  sendtextnote  Send a text note to the relay
    args: <content> - String of the text sent to the relay
  setmetadata   Set the metadata for this client
    args: <username> <about> <picture> - The client's metadata strings to set
  recommendserver       Recommend a server to the relay
    args: <urlrelay> - The url string of the server to recommend
  opensubscription      Open a subscription to the relay
    args: <subscriptionid> <kinds> <since> <until> - The metadata strings of the subscription to open
  closesubscription     Close a subscription to the relay
    args: <subscriptionid> - The subscription to close
  shutdown      Shutdown the relay
    args: None
```

**Suggestion** for taking it to the next level:
As we already added the `clap` dependency with https://github.com/civkit/civkit-node/pull/12, it might make sense to also use it here for a more future-proof result alongside the added capabilities of `clap`.

Towards that direction, those are some **Resources** that I found that we could maybe modify to our needs:
- `clap` has an [example for creating a repl](https://github.com/clap-rs/clap/blob/master/examples/repl.rs). 
- code from another project that uses `clap` to [to create a repl](https://github.com/dimfeld/perceive/blob/master/crates/perceive-cli/repl.rs).
- Some other implementations that I found in the wild also use [rustyline](https://github.com/kkawakam/rustyline) alongside clap, and in fact[ `clap-repl `](https://lib.rs/crates/clap-repl) combines both.

I am not sure what the sentiment is towards increasing the number of dependencies, but at least for `civkit-sample` (not target for production env) it seems like a reasonable choice.

**Comment by vladimirfomene** [comment]
@kouloumos, I like the idea of enhancement this change with `clap` and I also think is okay to add dependencies here given that the sample is mostly for testing and dev. I wonder what are @ariard thoughts around this.

**Comment by ariard** [comment]
> I like the idea of enhancement this change with clap and I also think is okay to add dependencies here given that the sample is mostly for testing and dev. I wonder what are @ariard thoughts around this.

One 100% here, we’ve already `clap` as a dependency (generally better to be conservative with dependencies though that’s okay for now given project maturity).

Feel free to add a “help” command for the `civkitd` / `civkit-cli` if you would like though all good to take the PR as it is with the `clap` integration.

File: src/sample.rs (hunk 1 above)
Comment (ariard) [reply src/sample.rs:61]: Yes the `shutdown` is shutting down the [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) that `civkit-sample` aims to.

Note, the design goal of the sample is to stay _relatively_ simple as a tool for developers and beta testers.

**Comment by vladimirfomene** [comment]
@ariard @kouloumos  this is ready for review. There is no need to implement help for `civkitd/civkit-cli` as Clap automatically implement the help command for you. While working on this I mistakenly called `cargo fmt` on my change. Tried undoing it but there are still some traces. I wonder what you guys think of integrating code formatting with `cargo fmt` as part of the contribution flow.

File: src/sample.rs [hunk 2]
```diff
@@ -27,173 +29,285 @@ use std::str::FromStr;
...
+    ";
+
+    Command::new("Nostr Client REPL")
+        .multicall(true)
+        .arg_required_else_help(true)
+        .subcommand_required(true)
+        .help_template(PARSER_TEMPLATE)
+        .subcommand(
+            Command::new("sendtextnote")
+                .args([arg!(-c --content <CONTENT> "Send a text note to the relay")])
```
Comment (kouloumos) [inline src/sample.rs:82]: This adds `content` as an option to the `sendtextnote` which results to the need of always prefixing the content with a flag. This pattern doesn't feel intuitive and makes the repl commands more complicate as `content` should be an argument of the command and not an option. Based on the [documentation of `Command`](https://docs.rs/clap/latest/clap/struct.Command.html) the right way to do it is using `Arg::new`: 
```suggestion
                .args([Arg::new("content").help("the content of the text note").required(true)])
```

This change also removes the need for checking if `content.is_none()` when matching because you can define the argument as required.

I believe that this comment applies to all the available commands of the repl; the `arg!(...)` pattern can change to `Arg::new(...).help(...).required(true)` to actually define arguments instead of options and simultaneously remove the need for args existence checking inside the individual commands. You can verify that this is indeed a change of behavior because of the need to add the flags in a74947c5d001b7cdf91fab18a67cd1771fbe2cb5.

File: src/sample.rs [hunk 3]
```diff
@@ -27,173 +29,285 @@ use std::str::FromStr;
...
+                .about("Shutdown the REPL"),
+        )
+}
+
+fn respond(
+    line: &str,
+    tx: &futures_channel::mpsc::UnboundedSender<Message>,
+    client_keys: &Keys,
+) -> Result<bool, String> {
+    let args = shlex::split(&line).expect("error: Failed to parse command");
```
Comment (kouloumos) [inline src/sample.rs:131]: I am not completely aware of the edge cases that this currently covers, but we could avoid the need for this extra dependency by keep using `.split_whitespace()`. I've done a few tests without noticing any issues.
```suggestion
    let args = line.split_whitespace().collect::<Vec<&str>>();
```

File: src/sample.rs [hunk 4]
```diff
@@ -27,173 +29,285 @@ use std::str::FromStr;
...
+        .map_err(|e| e.to_string())?;
+    let mut words = args.iter();
+    let _ = words.next();
+
+    match matches.subcommand() {
+        Some(("sendtextnote", matches)) => {
+            let content: Option<&String> = matches.get_one("content");
+            if content.is_none() {
+                return Err(
+                    "ERROR: sendtextnote has 1 required argument: `sendtextnote content`"
```
Comment (kouloumos) [inline src/sample.rs:143]: After switching `content` from an argument to an option, these error messages are now misleading.

**Review by kouloumos — COMMENTED** [review]
The integration looks good, I left some comments inline.

I think d4c6aeb588a68c88deb3b66a30338e088aaca2d5 can be dropped, it implements something that is reverted in later commits.

File: src/sample.rs [hunk 5]
```diff
@@ -27,173 +29,285 @@ use std::str::FromStr;
...
+    line: &str,
+    tx: &futures_channel::mpsc::UnboundedSender<Message>,
+    client_keys: &Keys,
+) -> Result<bool, String> {
+    let args = shlex::split(&line).expect("error: Failed to parse command");
+    let matches = cli()
+        .try_get_matches_from(args.clone())
+        .map_err(|e| e.to_string())?;
+    let mut words = args.iter();
+    let _ = words.next();
```
[the remark this answers is not in the document]
Comment (vladimirfomene) [reply src/sample.rs:136]: It has no use. Sorry I forgot to clean that up.

File: src/sample.rs (hunk 2 above)
Comment (vladimirfomene) [reply src/sample.rs:82]: Makes sense!

File: src/sample.rs (hunk 3 above)
Comment (vladimirfomene) [reply src/sample.rs:131]: Makes sense

**Comment by vladimirfomene** [comment]
@kouloumos thanks for the review.

File: src/sample.rs [hunk 6]
```diff
@@ -27,173 +29,258 @@ use std::str::FromStr;
...
+        .subcommand(
+            Command::new("sendtextnote")
+                .args([Arg::new("content").help("the content of the text note").required(true)])
+                .help_template(APPLET_TEMPLATE)
+                .about("Send a text note to the relay"),
+        )
+        .subcommand(
+            Command::new("setmetadata")
+                .args([
+					Arg::new("username").help("The client's username").required(true),
```
Comment (ariard) [inline src/sample.rs:89]: could be aligned though nevermind

we could add the same coding conventions than LDK’s one here: https://github.com/lightningdevkit/rust-lightning/blob/main/CONTRIBUTING.md#coding-conventions to be honest for now i’m quite shrug about strong rust coding conventions

**Review by ariard — APPROVED** [review]
Code review and tested ACK 3e83fb35c5, thanks for the cleanup!

**Comment by ariard** [comment]
> I wonder what you guys think of integrating code formatting with cargo fmt as part of the contribution flow.

We can use `clippy` in the same way done by LDK. For code conventions, they’re documented in LDK’s CONTRIBUTING.md, from my memory they’re not exactly the same than `rustfmt` because some shared with rust bitcoin libs for historical reasons. As a rule of thumb, I think let’s just do the same than LDK in matters of code conventions and contribution flow, all proportion guarded in term of project maturity.

**Comment by vladimirfomene** [comment]
By the way, `rust-bitcoin` uses `rustfmt` unlike `LDK`. See here: https://github.com/rust-bitcoin/rust-bitcoin/blob/master/CONTRIBUTING.md#coding-conventions. Personally, I'm thinking we should just use the default rust formatting option. I'm okay if we go with no LDK code convention though. I have no strong feelings about this

**Comment by ariard** [comment]
> By the way, rust-bitcoin uses rustfmt unlike LDK. See here: https://github.com/rust-bitcoin/rust-bitcoin/blob/master/CONTRIBUTING.md#coding-conventions.

Okay thanks for the reminder, I remembered we did something non-conventional for `LDK`. If we think using `rustfmt` will help the workflow of new contributors, especially ones without a strong Rust background I think we can go for it.

From `rust-bitcoin`, "The repository currently uses rustfmt (WIP, some directories are excluded).” We can do the same and exclude directories or files and case-by-case at the CI level if we’re not happy with it.
